ghc-lib-gen/src/Main.hs

```diff
@@ -96,205 +127,52 @@ extraFiles =
-  , "GhcMonad"
-  , "GhcPrelude"
-  , "HaddockUtils"
-  , "HeaderInfo"
-  , "Hooks"
-  , "HsBinds"
-  , "HsDecls"
-  , "HsDoc"
-  , "HsExpr"
-  , "HsExtension"
-  , "HsImpExp"
-  , "HsInstances"
-  , "HsLit"
-  , "HsPat"
-  , "HsSyn"
-  , "HsTypes"
-  , "HsUtils"
-  , "HscTypes"
-  , "IOEnv"
-  , "Id"
-  , "IdInfo"
-  , "IfaceSyn"
-  , "IfaceType"
-  , "InstEnv"
-  , "InteractiveEvalTypes"
-  , "Json"
-  , "Kind"
-  , "KnownUniques"
-  , "Language.Haskell.TH"
-  , "Language.Haskell.TH.LanguageExtensions"
-  , "Language.Haskell.TH.Lib"
-  , "Language.Haskell.TH.Lib.Internal"
-  , "Language.Haskell.TH.Lib.Map"
-  , "Language.Haskell.TH.Ppr"
-  , "Language.Haskell.TH.PprLib"
-  , "Language.Haskell.TH.Syntax"
-  , "Lexeme"
-  , "Lexer"
-  , "LinkerTypes"
-  , "ListSetOps"
-  , "Literal"
-  , "Maybes"
-  , "MkCore"
-  , "MkId"
-  , "Module"
-  , "MonadUtils"
-  , "Name"
-  , "NameCache"
-  , "NameEnv"
-  , "NameSet"
-  , "OccName"
-  , "OccurAnal"
-  , "OptCoercion"
-  , "OrdList"
-  , "Outputable"
-  , "PackageConfig"
-  , "Packages"
-  , "Pair"
-  , "Panic"
-  , "Parser"
-  , "PatSyn"
-  , "PipelineMonad"
-  , "PlaceHolder"
-  , "PlainPanic"
-  , "Platform"
-  , "PlatformConstants"
-  , "Plugins"
-  , "PmExpr"
-  , "PprColour"
-  , "PprCore"
-  , "PrelNames"
-  , "PrelRules"
-  , "Pretty"
-  , "PrimOp"
-  , "RdrHsSyn"
-  , "RdrName"
-  , "RepType"
-  , "Rules"
-  , "Settings"
-  , "SizedSeq"
-  , "SrcLoc"
-  , "StringBuffer"
-  , "SysTools.BaseDir"
-  , "SysTools.Terminal"
-  , "TcEvidence"
-  , "TcRnTypes"
-  , "TcType"
-  , "ToIface"
-  , "ToolSettings"
-  , "TrieMap"
-  , "TyCoRep"
-  , "TyCon"
-  , "Type"
-  , "TysPrim"
-  , "TysWiredIn"
-  , "Unify"
-  , "UniqDFM"
-  , "UniqDSet"
-  , "UniqFM"
-  , "UniqSet"
-  , "UniqSupply"
-  , "Unique"
-  , "Util"
-  , "Var"
-  , "VarEnv"
-  , "VarSet"
-  ]
+-- | Not finished but the idea here is to calculate via `ghc -M` the
+-- list of modules that are required for 'ghc-lib-parser'.
+calcParserModules :: IO [String]
+calcParserModules = do
+  lib <- mapM readCabalFile cabalFileLibraries
+  let include_dirs = map ("-I" ++ ) ghcLibParserIncludeDirs
+      hs_source_dirs = map ("-i" ++ ) $ ghcLibParserHsSrcDirs lib
+      cmd = unwords $
+        [ "stack exec -- ghc"
+        , "-dep-suffix ''"
+        , "-dep-makefile .parser-depends"
+        , "-M"]
+        ++ include_dirs
+        ++ ["-ignore-package ghc-lib-parser -package ghc", "-package base"]
+        ++ hs_source_dirs ++ ["-ighc-lib/stage0/libraries/ghc-heap/build"]
+        ++ ["ghc-lib/stage0/compiler/build/Parser.hs"]
+  putStrLn "# Generating 'ghc/.parser-depends'..."
+  system_ cmd
 
--- | Stub out a couple of definitions in the ghc-heap library that require CMM features,
---   since Cabal doesn't work well with CMM files.
+  buf <- readFile' ".parser-depends"
+  let depends = filter (not . isPrefixOf "#") (lines buf)
+      depends' = filter (isSuffixOf ".hs") depends
+      srcPaths = map snd (mapMaybe (stripInfix ":") depends')
```

Comment:
This looks like a lot of repetition and duplication and it isn’t entirely clear what you are trying to accomplish with that.
I would suggest that you add a comment with an example of an input path and the final path after all this magic.
Presumably your goal here is to strip away the source dir? If so, should this be based on `ghcLibParserHsSrcDirs`?